File: 2022/Day 6/pt1.cpp
#include "helper.hpp"
#include <map>

#define SIZE 4

bool check_unique(std::string substring)
{
    std::map<char, int> char_map;
    for (char c : substring)
    {
        if (char_map[c] == 1)
            return false;
        char_map[c] = 1;
    }
    return true;
}

int get_start_of_packet(std::string line)
{
    for (int i = SIZE; i <= line.size(); i++)
    {
        std::string substring = line.substr(i - SIZE, SIZE);
        if (check_unique(substring))
            return i;
    }

    return -1;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");
    std::string line = lines[0];

    auto t1 = Helper::get_current_time();
    int start_of_packet = get_start_of_packet(line);
    std::cout << "Start of Packet: " << start_of_packet << std::endl;
    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/day_08/pt1.go
package day08

import (
	"fmt"
	"regexp"

	"github.com/benjababe/advent-of-code/helper"
)

func solvePt1(lines []string) int64 {
	score := int64(0)

	instructions := lines[0]
	nodeMap := make(map[string](map[rune]string))
	nodeRegex := regexp.MustCompile(`(\w+)\s=\s\((\w+),\s(\w+)\)`)

	for _, line := range lines[1:] {
		if len(line) == 0 {
			continue
		}

		match := nodeRegex.FindStringSubmatch(line)
		src, l, r := match[1], match[2], match[3]
		nodeMap[src] = map[rune]string{'L': l, 'R': r}

		helper.Unused(line, match)
	}

	curNode := "AAA"
	for curNode != "ZZZ" {
		for _, inst := range instructions {
			curNode = nodeMap[curNode][inst]
			score++
		}
	}

	return score
}

func Pt1() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solvePt1(lines)
	fmt.Printf("Day 8\tPt1:\t%d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)
}
